File: Demo/Sources/ViewControllers/DataView/Calendar/CalendarDataViewController.m
//
//  Globus
//

#import "CalendarDataViewController.h"
#import "CalendarMonthDataViewCell.h"
#import "CalendarWeekdayDataViewCell.h"
#import "CalendarDayDataViewCell.h"

@interface CalendarDataViewController () {
    GLBDataViewCalendarContainer* _dataViewContainer;
}

@end

@implementation CalendarDataViewController

#pragma mark - Init / Free

- (void)setup {
    [super setup];
    
    self.edgesForExtendedLayout = UIRectEdgeNone;
    self.title = @"CalendarDataView";
}

#pragma mark - UIViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    [self.navigationItem glb_removeAllLeftBarButtonItemsAnimated:NO];
    [self.navigationItem glb_addLeftBarButtonNormalImage:[UIImage imageNamed:@"MenuButton"] target:self action:@selector(pressedMenu:) animated:NO];
}

#pragma mark - GLBViewController

- (void)update {
    [super update];
    
    [self.dataView batchUpdate:^{
        NSDate* now = NSDate.date;
        [_dataViewContainer prepareBeginDate:now.glb_beginningOfMonth endDate:now.glb_endOfMonth];
    }];
}

- (void)clear {
    [self.dataView batchUpdate:^{
        [_dataViewContainer cleanup];
    }];
    
    [super clear];
}

#pragma mark - GLBDataViewController

- (void)configureDataView {
    _dataViewContainer = [GLBDataViewCalendarContainer containerWithCalendar:NSCalendar.currentCalendar];
    
    [self.dataView registerIdentifier:GLBDataViewCalendarMonthIdentifier withViewClass:CalendarMonthDataViewCell.class];
    [self.dataView registerIdentifier:GLBDataViewCalendarWeekdayIdentifier withViewClass:CalendarWeekdayDataViewCell.class];
    [self.dataView registerIdentifier:GLBDataViewCalendarDayIdentifier withViewClass:CalendarDayDataViewCell.class];
    
    self.dataView.container = _dataViewContainer;
}

- (void)cleanupDataView {
    [super cleanupDataView];
    
    _dataViewContainer = nil;
}

#pragma mark - Actions

- (IBAction)pressedMenu:(id)sender {
    [self.glb_slideViewController showLeftViewControllerAnimated:YES complete:nil];
}

@end

File: Demo/Sources/ViewControllers/DataView/Calendar/CalendarDataViewController.swift
//
//  Globus
//

import Globus

class CalendarDataViewController: GLBDataViewController {
    
    // MARK: - Property
    
    public var dataViewContainer: GLBDataViewCalendarContainer?
    
    // MARK: - Init / Free
    
    override func setup() {
        super.setup()
        
        self.edgesForExtendedLayout = UIRectEdge(rawValue: 0);
        self.title = "CalendarDataView"
    }
    
    // MARK: - UIViewController
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.navigationItem.glb_removeAllLeftBarButtonItems(animated: false)
        self.navigationItem.glb_addLeftBarButtonNormalImage(UIImage(named: "MenuButton"), target: self, action: #selector(pressedMenu(_ :)), animated: false)
    }
    
    // MARK: - GLBViewController
    
    override func update() {
        super.update()
        
        self.dataView?.batchUpdate({
            let now = NSDate()
            self.dataViewContainer?.prepareBegin(now.glb_beginningOfMonth()!, end: now.glb_endOfMonth()!)
        })
    }
    
    override func clear() {
        self.dataView?.batchUpdate({
            self.dataViewContainer?.cleanup()
        })
        
        super.clear()
    }
    
    // MARK: - GLBDataViewController
    
    override func configureDataView() {
        self.dataViewContainer = GLBDataViewCalendarContainer(calendar: Calendar.current)
        
        self.dataView?.registerIdentifier(GLBDataViewCalendarMonthIdentifier, withViewClass: CalendarMonthDataViewCell.self)
        self.dataView?.registerIdentifier(GLBDataViewCalendarWeekdayIdentifier, withViewClass: CalendarWeekdayDataViewCell.self)
        self.dataView?.registerIdentifier(GLBDataViewCalendarDayIdentifier, withViewClass: CalendarDayDataViewCell.self)
        
        self.dataView?.container = self.dataViewContainer;
    }
    
    override func cleanupDataView() {
        super.cleanupDataView();
        
        self.dataViewContainer = nil;
    }
    
    // MARK: - Actions
    
    @IBAction internal func pressedMenu(_ sender: Any) {
        self.glb_slideViewController?.showLeftViewController(animated: true, complete: nil)
    }
    
}
